cli/internal/runsummary/spaces.go

```diff
@@ -1,13 +1,235 @@
 package runsummary
 
 import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/mitchellh/cli"
 	"github.com/vercel/turbo/cli/internal/ci"
+	"github.com/vercel/turbo/cli/internal/client"
 )
 
-// spacesRunResponse deserialized the response from POST Run endpoint
-type spacesRunResponse struct {
-	ID  string
-	URL string
+const runsEndpoint = "/v0/spaces/%s/runs"
+const runsPatchEndpoint = "/v0/spaces/%s/runs/%s"
+const tasksEndpoint = "/v0/spaces/%s/runs/%s/tasks"
+
+// spaceRequest contains all the information for a single request to Spaces
+type spaceRequest struct {
+	method  string
+	url     string
+	body    interface{}
+	makeURL func(self *spaceRequest, r *spaceRun) error // Should set url on self
+	onDone  func(self *spaceRequest, response []byte)   // Handler for when request completes
+}
+
+func (req *spaceRequest) error(msg string) error {
+	return fmt.Errorf("[%s] %s: %s", req.method, req.url, msg)
+}
+
+type spacesClient struct {
+	rsm      *Meta
+	requests chan *spaceRequest
+	errors   []error
+	api      *client.APIClient
+	ui       cli.Ui
+	run      *spaceRun
+	wg       sync.WaitGroup
+	spaceID  string
+	enabled  bool
+}
+
+type spaceRun struct {
+	ID      string
+	URL     string
+	created chan struct{} // a signal that the run has completed
+}
+
+func newSpacesClient(spaceID string, api *client.APIClient, ui cli.Ui, rsm *Meta) *spacesClient {
+	return &spacesClient{
+		api:      api,
+		ui:       ui,
+		rsm:      rsm,
+		spaceID:  spaceID,
+		enabled:  spaceID != "",
+		requests: make(chan *spaceRequest), // TODO: give this a size based on tasks
```

Comment:
Without a buffer here, you will block execution once you have request in flight equal to the number of processing goroutines you have. That might be ok, depending on what you are doing in the processing goroutines (See below)